docs: add block specification in data access (#84)
RequestNetwork/requestNetwork: packages/data-access/specs/request-block-specification-v0.1.0.md

# Request Block Specification v0.1.0

You can be interested in this document if:

- You want to create your own implementation of the request protocol
- You are curious enough to dive and see what is under the hood of the request protocol

You don't need to read this document if:

- You want to develop an app using the request protocol (see the API library instead [here](/packages/request-client.js))

The Data Access layer is responsible for:

- Indexing transactions to allow retrieval

> - Batching transactions

Reviewer: data-access is also responsible to synchronize with storage and manage the local cache

Author: This is not in the scope of the request block

Reviewer 2: it can be clearer with an introduction explaining what this document is about

## Transactions indexing (Channel & Topics)

> The indexing of the transactions is made by two mechanisms:

Reviewer 2: ```suggestion
The indexing of the transactions is performed through two mechanisms:
```

- Channel
- Topics

> The channel in data-access allows to group several transactions under one string: the channel id. This channel is indexed by several strings called topics.

Reviewer: ```suggestion
The channel in data-access allows to group several transactions under one string: the channel id. This channel can eventually be indexed by several strings called topics.
```

> The channel in data-access allows to group several transactions under one string: the channel id. This channel can be indexed by several strings called topics.

Reviewer 2: to add context, we could explain why it's useful to have channels and topics

There are two ways of getting transactions:

- by channel id: this gets all the transactions of the channel
- by topic: this gets all the transactions of the channels indexed by this topic

## Transaction batching

> The transaction are stored and indexed in blocks.

Reviewer: ```suggestion
The transactions are stored and indexed in blocks.
```

> The transactions are stored and indexed in blocks.

Reviewer 3: missing a definition of a block?

Author: I give it just after

Reviewer 3: hum, it defines the structure of a block, not what's the definition of a block...
I mention this because I had (and still have) a hard time with this word, it's very vague in the protocol concept imo

Reviewer 2: explaining the why should make it much clearer for the reader

> A block is JSON object containing two main parts:

Reviewer: ```suggestion
A block is a JSON object containing two main parts:
```

- header: containing the information to index the transactions (channels and topics)
- transactions: the transactions themselves

## Block Properties

| Property              | Type   | Description                                                     |
| --------------------- | ------ | --------------------------------------------------------------- |

> | **header.channelIds** | Object | Array of positions in transactions Array indexed by channel ids |

Reviewer: ```suggestion
| **header.channelIds** | Object | Array of positions in the array of transactions indexed by channel ids |
```

> | **header.topics**     | Object | Array of topics indexed by channel ids                          |

Reviewer: Mhh, it's rather the channel id that is indexed by a topic not the opposite, isn't it ?

Author: nope it is not :)
